src/curvefit/initializer/initializer.py

```diff
@@ -0,0 +1,352 @@
+    has priors that are informed by whatever `PriorInitializerComponent`s were passed in.
+
+    These PriorInitializerComponents are **model-specification-specific**. You need to be aware of what
+    component you are passing in and that it matches your model_prototype. If it doesn't, an error will be thrown.
+    The requirements are listed in the description
+    sections of each [`PriorInitializerComponent`](PriorInitializerComponent.md).
+
+    ## Arguments
+
+    - `self.prior_initializer_components (List[PriorInitializerComponent])` a list of prior initializer
+        components (instantiated) to use in updating the parameter set
+
+    ## Attributes
+
+    - `self.component_types (List[str])`: list of the types of initializer components
+    - `self.joint_solver (curvefit.solvers.solver.Solver)`: a solver/model run on all of the data
+    - `self.individual_solvers (List[curvefit.solvers.solver.Solver]): a list of solver/models run on each
+        group in the data individually
+
+    ## Methods
+
+    ### `initialize`
+    For a given data, model prototype and solver prototype, run the prior initialization for all
+    prior initializer components and return an updated parameter set.
+
+    - `data (curvefit.core.data.Data)`: a Data object that represents all of the data that will be
+        used in the initialization (this will often be a subset of all available data)
+    - `model_prototype (curvefit.core.model.CoreModel)`: a model that will be used as the prototype
+        and cloned for the joint and individual fits
+    - `solver_prototype (curvefit.solvers.solver.Solver): a solver that will be used as the prototype
+        and cloned for the joint and individual fits
+
+    ## Usage
+
+    ```python
+    prior_initializer = PriorInitializer([LnAlphaBetaPrior(), BetaPrior()])
+    new_parameter_set = prior_initializer.initialize(
+        data=data, model_prototype=model_prototype,
+        solver_prototype=solver_prototype, parameter_set_prototype
+    )
+    ```
+
+    {end_markdown PriorInitializer}
+    """
+    def __init__(self, prior_initializer_components):
+
+        self.prior_initializer_components = prior_initializer_components
+        self.component_types = [c.component_type for c in self.prior_initializer_components]
+
+        self.joint_solver = None
+        self.individual_solvers = None
+
+    @staticmethod
+    def _run_joint(data, model_prototype, solver_prototype):
+        model = model_prototype.clone()
+        solver = solver_prototype.clone()
+
+        solver.set_model_instance(model)
+        solver.fit(data=data)
+        return solver
+
+    @staticmethod
+    def _run_individual(data, model_prototype, solver_prototype):
+        solvers = list()
+
+        for group in data.groups:
+            model = model_prototype.clone()
+            solver = solver_prototype.cone()
+
+            group_data = data.get_df(group=group, copy=False, return_specs=True)
```

Comment:
looks great! I think this may supersede #78, the smart initialization piece? what do you think @mbannick @zhengp0 ? also I just have a question regarding `clone()`. was the purpose of cloning to preserve data stored in it? (e.g. for `model` the `data_inputs` and for `solver` the `x_opt` and `fun_val_opt`.)

Reply:
Thanks! For `clone()` it is actually because we will need to pass instances of classes to the eventual ModelRun function rather than just the class itself, and we will need a lot of copies of that instance because we will run many different models (each group / each data point in PV), and also here for initialization

Reply:
For example, someone will pass
```
run = ModelRun(model=CoreModel(param_set=ParameterSet(...), curve_fun=..., loss_fun=...)
```
and we will need to make copies of that instance of `CoreModel` to preserve the `curve_fun` and `loss_fun` attributes.

Reply:
@jizezhang

Reply:
yes we would definitely need to pass in instances. I was actually just wondering about the `for` loop starting at line 307. It seems that nothing about the model changed except the data part, so I was wondering if we could just save data (which can be detached from model) instead of copying the model? (just trying to see if we could copy less)

Reply:
also for the solver

Reply:
Ah, I think `ModelPrototype` won't have data attached to it at this point at all! so it's literally just copying the skeleton

Reply:
same with the solver

Reply:
ah, yes indeed. for solvers it would just be the skeleton. for models when `fit` is called, it would attach data as `data_inputs` to the model. we could remove that via `erase_data` or `detach_data` (which will return the `data_inputs`, if we just want the skeleton.

Reply:
I think I get what you meant now! ignore what I said above!